### ISSUE TITLE
Blender Python API bindings generation issues

### ISSUE THREAD

**Issue by liamcurry** [AndrejOrsula/pyo3_bindgen#23]
Hi @AndrejOrsula,

First of all thank you for creating this library, this is exactly what I was looking for and I was happy to see you had already started it.

I am starting to write Blender addons with Rust, and since Blender doesn't seem to support C FFI it appears that calls need to go through their Python API. Here are their docs: https://docs.blender.org/api/current/index.html

I have the basics of this working, where I am able to make changes in Rust and see them hot-reloaded in Blender.

I am using `pyo3_bindgen` to generate the Python bindings that I can use in Rust, which would otherwise be a very long and tedious task.

The Python package I am trying to generate bindings for is `bpy`: https://pypi.org/project/bpy/

This can be installed as just a library without having Blender installed.

I am seeing a few different issues when generating bindings:

1. Running `pyo3_bindgen --module-names bpy` does not generate bindings recursively for the whole module as I expected, maybe this is intended behavior?
2. The `bpy.ops` module is never populated, the bindings are just an empty Rust module https://docs.blender.org/api/current/bpy.ops.html
3. The `bpy.types` module is absolutely massive, currently at **499,754 lines of code**. This is so large that it basically makes the whole project unusable if I try to include it.

One thing I noticed from looking at the generated code is that it always uses the full path for things needed in `pyo3`, for example:

```rust
    pub fn set_app<'py>(
        py: ::pyo3::marker::Python<'py>,
        p_value: impl ::pyo3::IntoPy<::pyo3::Py<::pyo3::types::PyTuple>>,
    ) -> ::pyo3::PyResult<()> {
        let p_value = ::pyo3::IntoPy::<::pyo3::Py<::pyo3::types::PyTuple>>::into_py(p_value, py);
        let p_value = p_value.bind(py);
        ::pyo3::types::PyAnyMethods::setattr(
            py.import_bound(::pyo3::intern!(py, "bpy"))?.as_any(),
            ::pyo3::intern!(py, "app"),
            p_value,
        )
    }
```

Would it be possible to have an option to import `pyo3` types so that the full path isn't used? The example above could then be condensed down to something like this:

```rust
    pub fn set_app<'py>(py: Python<'py>, p_value: impl IntoPy<Py<PyTuple>>) -> PyResult<()> {
        let p_value = IntoPy::<Py<PyTuple>>::into_py(p_value, py);
        let p_value = p_value.bind(py);
        PyAnyMethods::setattr(
            py.import_bound(intern!(py, "bpy"))?.as_any(),
            intern!(py, "app"),
            p_value,
        )
    }
```

It seems like a relatively small change but at the scale of the ~500k `bpy.types` module I think it would make a significant impact. Still, even if it were ~300k I'm not sure it would be very usable as a single module. Maybe this is a case where generated bindings are not the right solution and something more custom is needed.

**Comment by AndrejOrsula** [comment]
Hey @liamcurry,

> Running `pyo3_bindgen --module-names bpy` does not generate bindings recursively for the whole module as I expected, maybe this is intended behavior?

TLDR: The Python API of Blender (`bpy`) is a very special case. You could try using [`blr`](https://github.com/AndrejOrsula/blr) and see if it works for you.

Thanks for opening this issue. Indeed, the Python API of Blender is a special case because `bpy` contains bindings to the underlying C/C++ implementation with its structure not being handled in the same way as other native Python modules. Furthermore, some members (`bpy.data`) are determined at runtime. There are also differences when Blender is compiled into `bpy` as opposed to the full application. This would require a special way of handling the discovery for the underlying submodules and members.

I tried with `bpy==3.6` a long time ago, but I did not investigate it further, seeing that I encountered the same limitation of `bpy.ops` submodules not being automatically discovered from within the shared FFI object. It wasn't my priority because I already made `blr` a while before starting this project (described below). It should technically be possible to use `pyo3_bindgen` for it, but it would require further investigation (it could be either something very minor or complex, but I have no idea at this point).

[`blr`](https://github.com/AndrejOrsula/blr) manually specifies the signatures of Blender members for which to generate bindings via declarative macros inside [`pyo3_more`](https://github.com/AndrejOrsula/pyo3_more), and it also has a more idiomatic interface for some aspects. It should contain a large chunk of the interfaces — and it is rather trivial to add what it missing with the macros. As it uses fully declarative macros, the compile times are also shorter this way. There are not many examples, as I am probably the only person who used it, but [`blr_procgen`](https://github.com/AndrejOrsula/blr_procgen) contains some procedural generation pipelines for [`drl_omni_peg`](https://github.com/AndrejOrsula/drl_omni_peg) using Geometry Nodes. It is based on 3.6 LTS, but it could be updated to 4.2 LTS if there is ever any interest in it, and combined with some autogeneration via `pyo_bindgen` once this issue is resolved.

One thing you need to keep in mind is that the Python API of Blender is not thread-safe. All generated functions are inherently `unsafe` for this reason (but not marked as `unsafe` to not make it too unergonomic). Only the main entry-point [`BlenderProject`](https://github.com/AndrejOrsula/blr/blob/0cc2d032c717e3c3ba098a27b368ae2626681df1/blr/src/project.rs#L47) struct includes logic for thread-safety.

---

> Would it be possible to have an option to import pyo3 types so that the full path isn't used?

This could result in identifier collision in cases where the Python module already contains a member with the same name (e.g. `PyTuple`). Sure, the option could be added, but it might result in a large development overhead and a lot of complexity in this codebase.

This has actually happened to me quite often while working on this crate, so I just decided to fully specify every type. As far as I understand, this is the preferred way for macros as it avoids any ambiguity — and the generated code is not really intended to be human-readable. This is my first time using proc macros, so I might be missing something.

I also know very little about compilers, but I have a feeling that either option wouldn't really affect the compilation times in any significant way. Perhaps the full path is even more performant as it simplifies the compiler's job of mapping the identifiers? However, the memory footprint might be higher due to the larger source code size.

---

If you want to decrease the volume of generated code, you can also generate bindings only for specific submodules, e.g. `pyo3_bindgen -m gymnasium.envs gymnasium.core.spaces gymnasium.experimental.wrappers`. There is also [blocklist](https://github.com/AndrejOrsula/pyo3_bindgen/blob/64c64a5f8a0fd8233918673a6845c2939af700d9/pyo3_bindgen_engine/src/config.rs#L51) that you can configure.

**Comment by AndrejOrsula** [comment]
> write Blender addons with Rust

Oh, I missed this part. I did not attempt to write Blender addons with Rust, and it is very likely that I did not even expose this functionality in `blr`.
